river/neural_net/mlp.py

```diff
@@ -1,326 +1,416 @@
+        # pass DataFrames with different column orders without having to worry.
+        a[1] = X.values[:, np.argsort(X.columns.to_numpy())]
+
+        for i in range(2, self.n_layers + 1):
+            z[i] = np.dot(a[i - 1], self.w[i - 1]) + self.b[i - 1]
+            a[i] = self.activations[i - 2].apply(z[i])
+
+        return z, a
+
+    def _backward(self, z: dict, a: dict, y: pd.DataFrame):
+        """Execute a backward pass through the neural network, aka. backpropagation.
+
+        Parameters
+        ----------
+        z
+            The output of each layer.
+        a
+            The output of each activation function.
+        y
+            A DataFrame of shape (batch_size, n_targets).
+
+        """
+
+        y = y.values[:, np.argsort(y.columns.to_numpy())]
+
+        # Determine the partial derivative and delta for the output layer
+        y_pred = a[self.n_layers]
+        final_activation = self.activations[self.n_layers - 2]
+        delta = self.loss.gradient(y, y_pred) * final_activation.gradient(y_pred)
+        dw = np.dot(a[self.n_layers - 1].T, delta)
+
+        update_params = {self.n_layers - 1: (dw, delta)}
+
+        # Go through the layers in reverse order
+        for i in range(self.n_layers - 2, 0, -1):
+            delta = np.dot(delta, self.w[i + 1].T) * self.activations[i - 1].gradient(z[i + 1])
+            dw = np.dot(a[i].T, delta)
+            update_params[i] = (dw, delta)
+
+        # Update the parameters
+        for i, (dw, delta) in update_params.items():
+            self._optimizers[i, 0].step(w=self.w[i], g=dw)
+            self._optimizers[i, 1].step(w=self.b[i], g=np.mean(delta, axis=0))
+
+    def learn_many(self, X: pd.DataFrame, y: pd.DataFrame):
+        """Train the network.
+
+        Parameters
+        ----------
+        X
+            A DataFrame of shape (batch_size, n_features).
+        y
+            A DataFrame of shape (batch_size, n_targets).
+
+        """
+
+        # We expect y to be 2D, even if there is only one target to predict
+        if isinstance(y, pd.Series):
+            y = y.to_frame()
+
+        # The weights need initializing during the first call to partial_fit
+        if not hasattr(self, "w"):
+            self.w, self.b = xavier_init(
+                dims=(X.shape[1], *self.hidden_dims, y.shape[1]), seed=self.seed
+            )
+            self.features = X.columns.to_numpy()
+            self.features.sort()
+            self.targets = y.columns.to_numpy()
+
+        z, a = self._forward(X)
+        self._backward(z, a, y)
+
+        return self
+
+    def __call__(self, X: pd.DataFrame):
+        """Make predictions.
+
+        Parameters
+        ----------
+        X
+            A DataFrame of shape (batch_size, n_features).
+
+        """
+        _, a = self._forward(X)
+        y_pred = a[self.n_layers]
+        return pd.DataFrame(y_pred, columns=self.targets, index=X.index)
+
+
+class MLPRegressor(base.Regressor, MLP):

```

Comment:
Is there any reason GitHub displays this code as green (new)? What have you changed to `MLP` and `MLPRegressor`?

Reply:
No, I had copy pasted the whole mlp script from another local folder. I've only added the MLPBinaryClassifier code.